ThesesFR.js

```diff
@@ -79,24 +85,80 @@ function doWeb(doc, url) {
 }
 
 function scrape(doc, url) {
-	var translator = Zotero.loadTranslator('web');
-	translator.setTranslator('951c027d-74ac-47d4-a107-9c3069ab7b48'); // https://github.com/zotero/translators/blob/master/Embedded%20Metadata.js
-	translator.setDocument(doc);
-	translator.setHandler('itemDone', function (obj, item) {
-		// add Tags
-		var tags = ZU.xpath(doc, '//span[contains(@property, "dc:subject")]');
-		if (tags.length > 0) {
-			item.tags = [];
-			for (let tag of tags) {
-				item.tags.push(tag.textContent.trim());
-			}
+	const xmlDocumentUrl = `${url}.rdf`;
+	
+	// Each thesis record has an underlying .rdf file
+	Zotero.Utilities.HTTP.doGet(xmlDocumentUrl, function (text) {
+		const parser = new DOMParser();
+		const xmlDoc = parser.parseFromString(text, 'application/xml');
+
+		// Skiping invalid or empty RDF files : prevents crashes while importing multiple records
+		if (xmlDoc.getElementsByTagName('parsererror')[0] || xmlDoc.children[0].childElementCount === 0) {
+			Z.debug('Skiping: Invalid or empty RDF file.');
+			return;
 		}
-			
-		item.complete();
-	});
-	translator.getTranslatorObject(function (trans) {
-		trans.itemType = "thesis";
-		trans.doWeb(doc, url);
+		
+		// Importing XML namespaces for parsing purposes
+		const ns = {
+			bibo: 'http://purorg/ontology/bibo/',
+			dc: 'http://purl.org/dc/elements/1.1/',
+			dcterms: 'http://purl.org/dc/terms/',
+			foaf: 'http://xmlns.com/foaf/0.1/',
+			marcrel: 'http://www.loc.gov/loc.terms/relators/',
+			rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#'
+		};
+	
+		const title = ZU.xpathText(xmlDoc, '//dc:title', ns);
+		
+		// The case should't exists : preve;nts crashes while importing multiple records
+		if (!title) return;
+
+		const newItem = new Zotero.Item();
+		newItem.itemType = 'thesis';
+		newItem.title = title;
+
+		ZU.xpath(xmlDoc, '//marcrel:aut//foaf:Person/foaf:name | //marcrel:dis//foaf:Person/foaf:name', ns).forEach((auth) => {
+			const author = ZU.cleanAuthor(auth.textContent, 'author', true);
+			newItem.creators.push(author);
+		});
+
+		// Supervisor(s) must be considered as contributor(s) for french thesis
+		ZU.xpath(xmlDoc, '//marcrel:ths//foaf:Person/foaf:name', ns).forEach((sup) => {
+			const supervisor = ZU.cleanAuthor(sup.textContent, 'contributor', true);
+			newItem.creators.push(supervisor);
+		});
+
+		const abstractNotes = ZU.xpath(xmlDoc, '//dcterms:abstract', ns);
+		newItem.abstractNote = abstractNotes.length > 0 ? abstractNotes[0].textContent : undefined;
```

Comment:
Why no just use `ZU.xpathText()` here? (and similarly in other cases below)

Reply:
My xpath skills was a bit rusty. I failed to get the first element of the abstract (and other nodes) with pure xpath.

Originally I tried to use this expression `//dcterms:abstract[1]` with `xpathText()` function, but this expression was wrong in my context. It should have been `(//dcterms:abstract)[1]`.

So I used a workaround by mixing xpath and javascript which is not the best implementation. Now that I have found the right xpath expression I'll fix the code to use `xpathText()` function.

Reply:
(Note that you can also use `text(docOrElem, selector[, index])` or `attr(docOrElem, selector, attribute[, index])` with CSS selectors instead of using XPaths.)

Reply:
I already fixed the code by using xpath (2c91144) and `Z.xpathText()` but that's a good trick to know.